### PR TITLE
[RTM] ENH: Erode TPM ROIs in T1 space, resample to BOLD

### LINKED ISSUE #652: Interface TCompCorRPT failed to run
As reported here https://neurostars.org/t/fmriprep-tcompcorr-crashes-on-empty-array/668, tCompCor gets an `IndexError: cannot do a non-empty take from an empty axes.`

### PR DESCRIPTION
**Pull request by effigies**
This PR significantly modifies the creation of CSF/WM masks and eroded BOLD masks to accommodate partial field-of-view (FoV) BOLD series. It does so by performing erosions in the higher-resolution T1 space, resampling to BOLD, and applying the BOLD mask to limit results to the BOLD FoV.

These last two steps are encapsulated in the new `_resample_and_mask` function, while the original `_tpm2roi` works in the T1 space.

Eroded BOLD
------------------
1) T1 skull-stripping mask (`t1_mask`) is eroded
2) `_resample_and_mask` ensures that the result is clipped by the BOLD mask, in case of partial FoV

fov_tcompcor_mask

Blue: T1 mask
Red: BOLD mask
White: Eroded mask

CSF/WM ROIs
--------------------
1) If T1 mask is eroded, apply to thresholded tissue probability map (TTPM)
2) Erode TTPM in T1 space
3) `_resample_and_mask` ensures that the result is clipped by the BOLD mask, in case of partial FoV

fov_csf_mask

Red: Eroded BOLD mask
White: CSF ROI mask

fov_wm_roi

Red: BOLD mask (not eroded)
White: WM ROI mask

Closes #652

### CODE REVIEW

File: fmriprep/interfaces/utils.py [hunk 1]
```diff
@@ -133,43 +135,55 @@ def _run_interface(self, runtime):
...
+    bold_mask_img = nb.load(bold_mask)
+
+    probability_map_data = (tpm_img.get_data() >= pthres).astype(np.uint8)
+
+    eroded_mask_file = bold_mask
+    if mask_erosion_mm > 0:
+        eroded_mask_file = os.path.abspath("eroded_mask.nii.gz")
+
+        mask_img = nb.load(t1_mask)
+        iter_n = int(mask_erosion_mm / max(mask_img.header.get_zooms()))
```
Comment (effigies) [inline fmriprep/interfaces/utils.py]: Time isn't really an issue, but I'll give the ball approach a shot. It seems reasonable.

**Comment by chrisfilo** [comment]
Would be worth running this on an old subject with full FOV to see if how
much the masks changed.

On Aug 10, 2017 8:28 AM, "Chris Markiewicz" <notifications@github.com>
wrote:

*@effigies* commented on this pull request.
------------------------------

In fmriprep/interfaces/utils.py
<https://github.com/poldracklab/fmriprep/pull/659#discussion_r132487027>:

> -        nb.Nifti1Image(epi_mask_data, epi_mask_nii.affine,
-                       epi_mask_nii.header).to_filename(eroded_mask_file)
-    else:
-        eroded_mask_file = epi_mask
-    probability_map_data[epi_mask_data != 1] = 0
+    tpm_img = nb.load(t1_tpm)
+    bold_mask_img = nb.load(bold_mask)
+
+    probability_map_data = (tpm_img.get_data() >= pthres).astype(np.uint8)
+
+    eroded_mask_file = bold_mask
+    if mask_erosion_mm > 0:
+        eroded_mask_file = os.path.abspath("eroded_mask.nii.gz")
+
+        mask_img = nb.load(t1_mask)
+        iter_n = int(mask_erosion_mm / max(mask_img.header.get_zooms()))

Time isn't really an issue, but I'll give the ball approach a shot. It
seems reasonable.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/poldracklab/fmriprep/pull/659#discussion_r132487027>,
or mute the thread

**Comment by effigies** [comment]
Tried on ds114/sub-02, with AFNI skull stripping and got an empty mask.

Turns out AFNI skull-stripping doesn't fill holes, so erosion works from inside-out as well as outside-in. To do this, we'd need to update `niworkflows` to add a fill-in step (AFNI has [3dmask_tool](https://afni.nimh.nih.gov/pub/dist/doc/program_help/3dmask_tool.html) and [3dinfill](https://afni.nimh.nih.gov/pub/dist/doc/program_help/3dinfill.html) that both seem like good candidates), so that we don't break \*CompCor for people using `--no-skull-strip-ants`.

Trying with ANTs, now...

**Comment by chrisfilo** [comment]
Might be worth opening another issue for this.

**Comment by oesteban** [comment]
FYI: https://github.com/poldracklab/mriqc/issues/552, https://github.com/poldracklab/mriqc/issues/554

**Comment by effigies** [comment]
ds114/sub-02, using the iterative approach...

aCompCor mask:
ds114_acompcor_mask

tCompCor mask:
ds114_tcompcor_mask

**Comment by chrisfilo** [comment]
LGTM!

File: fmriprep/interfaces/utils.py (hunk 1 above)
Comment (effigies) [inline fmriprep/interfaces/utils.py]: Making this change induced a `MemoryError`. I'll make a new issue to consider moving to that approach, but for now it doesn't seem worth the time to dig into.

**Pull request merged**